accepted/2020/mono-convergence/platform-specific-httpclient.md

```diff
@@ -0,0 +1,452 @@
+
+**Example**
+```c#
+// This should use the default message handler for the application. Either as
+// preselected by the framework or set in the Project Options for the project.
+
+HttpClient client = new HttpClient();
+
+// Same behavour should be observed when using following code
+HttpClient client = new HttpClient(new HttpClientHandler());
+```
+
+There are several reasons why using `SocketHttpHandler` is not desirable or not even possible on some platforms
+
+* Lack of platform native TLS support for better performance
+* Bloats the application size
+* Performance and power consumption impact of managed implementation
+* Integration with the platform settings
+* Unavailability of Socket APIs
+
+### Platform-specific Http properties need to be available for customization
+In cases where HTTP handler wraps platform-specific implementation and the implementation has a way to control HTTP connection behaviour or properties, such properties need to be exposed to C# developers on the public platform-specific type.  If HttpClientHandler itself exposes such a knob, it needs to control the associated customization on the underlying platform.
+
+A few examples of platform handlers control properties
+
+* [cachePolicy](https://developer.apple.com/documentation/foundation/nsurlrequest/cachepolicy/useprotocolcachepolicy)
+* [allowsConstrainedNetworkAccess](https://developer.apple.com/documentation/foundation/nsmutableurlrequest/3325676-allowsconstrainednetworkaccess)
+* [allowsExpensiveNetworkAccess](https://developer.apple.com/documentation/foundation/nsmutableurlrequest/3325677-allowsexpensivenetworkaccess)
+* [networkServiceType](https://developer.apple.com/documentation/foundation/nsmutableurlrequest/1412378-networkservicetype)
+* [ReadTimeout](https://docs.oracle.com/javase/7/docs/api/java/net/URLConnection.html#setReadTimeout(int))
+* [AllowUserInteraction](https://docs.oracle.com/javase/7/docs/api/java/net/URLConnection.html#setAllowUserInteraction(boolean))
+
+## Xamarin.iOS/tvOS/watchOS
+There are 3 handlers available where **NSUrlSession** handler is the default. Except for watchOS where only NSUrlSessionHandler is supported.
+
+The handler is only preselected as the default with [UI options available](https://docs.microsoft.com/en-us/xamarin/cross-platform/macios/http-stack#selecting-an-httpclient-stack) using VS Project Settings dialog to change them. This is required to “fix” existing HttpClient code which was written with the assumption that the default handler will be the one working best on the platform.
+
+The ILLinker is used to remove the unnecessary handlers from the final binary in this case.
+
+### Handlers in Detail
+**SocketsHttpHandler Handler**
+
+The handler which is purely managed and behaves generally the same across all platforms except WatchOS (sockets are not available). The downside of using it is mostly felt in performance compared to the native handlers.
```

Comment:
It's a managed implementation, but for a few very specific things, mainly related to getting proxy information from the OS, we do P/Invoke.